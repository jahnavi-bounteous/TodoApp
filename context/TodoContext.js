import React, { createContext, useReducer, useContext, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const ADD_TODO = 'ADD_TODO';
export const DELETE_TODO = 'DELETE_TODO';
export const EDIT_TODO = 'EDIT_TODO';
export const SET_TODOS = 'SET_TODOS';
export const TOGGLE_IMPT = 'TOGGLE_IMP';

const TodoContext = createContext();

export const useTodoContext = () => {
  return useContext(TodoContext);
};

const todoReducer = (state, action) => {
  switch (action.type) {
    case SET_TODOS:
      return action.payload;
    case ADD_TODO:
      return [...state, action.payload];
    case DELETE_TODO:
      return state.filter((todo) => todo.id !== action.payload);
    case EDIT_TODO:
      return state.map((todo) =>
        todo.id === action.payload.id ? action.payload : todo
      );
      case TOGGLE_IMPT:
        return state.map((todo) =>
        todo.id === action.payload ? { ...todo, isImportant: !todo.isImportant } : todo
      );
    default:
      return state;
  }
};

export const TodoProvider = ({ children }) => {
  const [todos, dispatch] = useReducer(todoReducer, []);

  useEffect(() => {
    const loadTodos = async () => {
      try {
        const storedTodos = await AsyncStorage.getItem('todos');
        if (storedTodos !== null) {
          dispatch({ type: SET_TODOS, payload: JSON.parse(storedTodos) });
        }
      } catch (error) {
        console.error('Error loading todos from AsyncStorage:', error);
      }
    };

    loadTodos();
  }, []);

  useEffect(() => {
    const saveTodos = async () => {
      try {
        await AsyncStorage.setItem('todos', JSON.stringify(todos));
      } catch (error) {
        console.error('Error saving todos to AsyncStorage:', error);
      }
    };

    saveTodos();
  }, [todos]);

  return (
    <TodoContext.Provider value={{ todos, dispatch }}>
      {children}
    </TodoContext.Provider>
  );
};