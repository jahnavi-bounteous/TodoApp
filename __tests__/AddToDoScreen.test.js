import React from 'react';
import renderer from 'react-test-renderer';
import AddTodoScreen from '../screens/AddToDoScreen';
import { TodoProvider } from '../context/TodoContext';

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));

// Mock moment
jest.mock('moment', () => () => ({
  format: (format) => '01/01/2023',
}));

describe('AddTodoScreen', () => {
  it('renders correctly', () => {
    const route = {
      params: {}
    };
    const tree = renderer
      .create(
        <TodoProvider>
          <AddTodoScreen route={route} />
        </TodoProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('sets todo value correctly', () => {
    const route = {
      params: {}
    };
    const component = renderer.create(
      <TodoProvider>
        <AddTodoScreen route={route} />
      </TodoProvider>
    );
    const instance = component.root;
    const input = instance.findByProps({ placeholder: 'Add Todo' });
    input.props.onChangeText('Test Todo');
    expect(input.props.value).toBe('Test Todo');
  });

  it('handles date selection correctly', () => {
    const route = {
      params: {}
    };
    const component = renderer.create(
      <TodoProvider>
        <AddTodoScreen route={route} />
      </TodoProvider>
    );
    const instance = component.root;
    const dateInput = instance.findByProps({ placeholder: 'Set date (optional)' });
    expect(dateInput.props.editable).toBeFalsy();
  });

  it('handles time selection correctly', () => {
    const route = {
      params: {}
    };
    const component = renderer.create(
      <TodoProvider>
        <AddTodoScreen route={route} />
      </TodoProvider>
    );
    const instance = component.root;
    const timeInput = instance.findByProps({ placeholder: 'Set time (optional)' });
    expect(timeInput.props.editable).toBeFalsy();
  });

//   it('calls handleSaveTodo function correctly', () => {
//     const route = {
//       params: {}
//     };
//     const navigation = {
//       goBack: jest.fn(),
//     };
//     const component = renderer.create(
//       <TodoProvider>
//         <AddTodoScreen navigation={navigation} route={route} />
//       </TodoProvider>
//     );
//     const instance = component.root;
//     const saveButton = instance.findByProps({ testID: 'saveButton' });
//     saveButton.props.onPress();
//     expect(navigation.goBack).toHaveBeenCalled();
//   });
});
