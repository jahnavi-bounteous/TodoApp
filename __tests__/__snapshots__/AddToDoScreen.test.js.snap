// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoScreen renders correctly 1`] = `
<View
  style={
    {
      "margin": 16,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Add Todo"
    style={
      {
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 2,
        "fontSize": 16,
        "marginBottom": 10,
        "padding": 16,
      }
    }
    value=""
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      editable={false}
      placeholder="Set date (optional)"
      style={
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderStyle": "solid",
          "borderWidth": 2,
          "flex": 1,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      editable={false}
      placeholder="Set time (optional)"
      style={
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderStyle": "solid",
          "borderWidth": 2,
          "flex": 1,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1e90ff",
        "borderRadius": 6,
        "marginVertical": 10,
        "opacity": 1,
        "paddingVertical": 12,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 3,
      }
    }
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      Add
    </Text>
  </View>
</View>
`;
